### PR TITLE
DB-9760 Make temporary tables local to their sessions

### PR DESCRIPTION
**Pull request by ascend1**
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

### CODE REVIEW

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/1101/

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/1104/

**Comment by ascend1** [comment]
I made a document to explain the details of this change: https://splicemachine.atlassian.net/wiki/spaces/PD/pages/1068204201/Local+temporary+tables

Hope it makes the reviewing easier.

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/QueryTreeNode.java [hunk 1]
```diff
@@ -1408,13 +1424,20 @@ protected final TableDescriptor getTableDescriptor(String tableName,
         if(schema.getUUID()==null)
             return null;
 
+        //If not in session schema, look for local temporary table first.
+        //If tables is not under SESSION schema, don't get it from LCC. Table version is not set there.
```
Comment (jyuanca) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/QueryTreeNode.java:1428]: Session schema is for storing global temporary tables.
http://db.apache.org/derby/docs/10.0/manuals/reference/sqlj33.html

However, it seems we never implement global temporary tables correctly. We may want to file a Jira and fix it.

**Comment by martinrupp** [comment]
is this only for master, no branch-2.8/3.0?

**Comment by ascend1** [comment]
> is this only for master, no branch-2.8/3.0?

It should be in 3.0 as well. I've been expecting quite some comments, so haven't back-ported it.
It's decided to not have this one in 2.8. Please see Michael's comment to the Jira ticket: https://splicemachine.atlassian.net/browse/DB-9760

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/GenericPreparedStatement.java [hunk 2]
```diff
@@ -905,7 +905,9 @@ public void completeCompile(StatementNode qt) throws StandardException {
                 updateMode = cursorInfo.updateMode;
             }
         }
-        isValid = true;
+        synchronized (this) {
+            isValid = true;
+        }
```
Comment (martinrupp) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/GenericPreparedStatement.java:910]: setValid() ?

**Review by martinrupp — APPROVED** [review]

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/GenericPreparedStatement.java (hunk 2 above)
Comment (OlegMazurov) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/GenericPreparedStatement.java:912]: Second that.

**Review by carolp-503 — APPROVED** [review]
+1

Please address/Resolve comments in a follow-on delivery

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/GenericPreparedStatement.java (hunk 2 above)
Comment (ascend1) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/GenericPreparedStatement.java]: Done.

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Pull request merged**